### DISCUSSION TITLE
Implementing a DAG graph with ak Arrays

### DISCUSSION THREAD

**Discussion by pankajp** [scikit-hep/awkward#2746] [Q&A] [answered]
I'm implementing a DAG with awkward arrays, storing operations (nodes) and data (edges) as two arrays in a numba structref based on the numba docs.
I am having trouble implementing graph traversal which works in numba). I'm even ready to accept implementing it in a obhmode fallback if that's the only possible way.
Here's the snippet I have till now.

```python
import awkward as ak
import numba as nb
import numpy as np
import weakref
from numba import njit
from numba.core import types
from numba.experimental import structref
import itertools
import collections

"""Define a DAG graph of operations and data.
Each operation can have 0 or more data as input and 0 or more data as output
Each data can have an optional input operation (which is generating the data), and multiple operations where it is required as input

Challange: Implement the Data.input() method which returns an Operation or None in numba mode, by hook or by crook (nb.objmode())
"""

# Define a StructRef.
# `structref.register` associates the type with the default data model.
# This will also install getters and setters to the fields of
# the StructRef.
@structref.register
class GraphType(types.StructRef):
    def preprocess_fields(self, fields):
        # This method is called by the type constructor for additional
        # preprocessing on the fields.
        # Here, we don't want the struct to take Literal types.
        return tuple((name, types.unliteral(typ)) for name, typ in fields)


# Define a Python type that can be use as a proxy to the StructRef
# allocated inside Numba. Users can construct the StructRef via
# the constructor for this type in python code and jit-code.

# The weakref cache dict to get back the graph from the operations and layers
# Note: Use of eval to access these to avoid pickling errors
_graph_cache = weakref.WeakValueDictionary()
_graph_types_cache = weakref.WeakKeyDictionary()
_graph_idx_seq = itertools.count()
class Graph(structref.StructRefProxy):
    def __new__(cls, ops, dat):
        graph_idx = next(_graph_idx_seq)
        ops = ak.Array(ops, with_name='Operation', behavior={
            'Operation':Operation,
            'graph': graph_idx,
        })
        dat = ak.Array(dat, with_name='Data', behavior={
            'Data': Data,
            'graph': graph_idx,
            ('__numba_typer__','Data','input', ()): Data._input__numba_typer__,
            ('__numba_lower__','Data','input', ()): Data._input__numba_lower__,
        })
        res = structref.StructRefProxy.__new__(cls, ops, dat)
        res.ops = ops
        res.dat = dat
        eval('_graph_cache')[graph_idx] = res
        eval('_graph_types_cache')[res] = nb.typeof(ops), nb.typeof(ops[0]), nb.typeof(dat), nb.typeof(dat[0])
        return res


# This associates the proxy with MyStructType for the given set of
# fields. Notice how we are not constraining the type of each field.
# Field types remain generic.
structref.define_proxy(Graph, GraphType, ["ops", "dat"])


class Operation(ak.Record):
    def graph(self):
        return get_graph_from_idx(self.behavior['graph'])
    def inputs(self):
        return [self.graph().dat[idx] for idx in self.inputs_idx]
    def outputs(self):
        return [self.graph().dat[idx] for idx in self.outputs_idx]

def get_graph_from_idx(idx):
    return eval('_graph_cache')[idx]

def _get_types(idx):
    return eval('_graph_types_cache')[eval('_graph_cache')[idx]]

class Data(ak.Record):
    def graph(self):
        return get_graph_from_idx(self.behavior['graph'])
    def input(self):
        print('input()', self)
        return self.graph().ops[self.input_idx] if self.input_idx >= 0 else None
    def outputs(self):
        return [self.graph().ops[idx] for idx in self.outputs_idx]

    def _input__numba_typer__(akt, extra=()):
        return nb.types.Optional(_get_types(akt.behavior['graph'])[1])()

    def _input__numba_lower__(context, builder, sig, args):
        print(f'_input__numba_lower__: sig.args={[type(a) for a in sig.args]}, sig.return_type={type(sig.return_type)}, {len(args)=}')
        ret_type = sig.return_type
        def input_lower(dat):
            # Fixme: TypingError: Failed in nopython mode pipeline
            print('get_node() for:', dat.idx)
            with nb.objmode(inp=ret_type):
                inp = dat.input()
            return inp
        return context.compile_internal(builder, input_lower, sig, args)

operations = ak.Array({
    'idx': range(3),
    'inputs_idx': [[], [0,1],[1,2]],
    'outputs_idx': [[1],[2,3],[]],
})
data = ak.Array({
    'idx':range(4),
    'input_idx': [-1, 0, 1, 1],
    'input_port': [-1, 0, 0, 1],
    'outputs_idx': [[1], [1,2], [2], []],
})
extra_op_params = {'prop1':np.sin(np.arange(3))}  # Arbitrary extra fields
for k, v in extra_op_params.items():
    operations[k] = v

g = Graph(operations, data)


@nb.njit()
def func(l, is_compiled=False):
    print('data:', l.idx, l)
    with nb.objmode():
        print('objmode input:', l.input())
    inp = l.input()
    if inp is None:
        print('inp:', 'is None')
    else:
        print('inp:', inp)
    print('Done:', is_compiled)
func.py_func(g.dat[1])
print('----------------')
func(g.dat[1], True)
```

I've run out of ideas to implement this. The error I'm getting is:
```
data: 1 {idx: 1, input_idx: 0, input_port: 0, outputs_idx: [1, 2]}
input() {idx: 1, input_idx: 0, input_port: 0, outputs_idx: [1, 2]}
objmode input: {idx: 0, inputs_idx: [], outputs_idx: [1], prop1: 0}
input() {idx: 1, input_idx: 0, input_port: 0, outputs_idx: [1, 2]}
inp: {idx: 0, inputs_idx: [], outputs_idx: [1], prop1: 0}
Done: False
----------------
_input__numba_lower__: sig.args=[<class 'awkward._connect.numba.arrayview.RecordViewType'>], sig.return_type=<class 'numba.core.types.misc.Optional'>, len(args)=1
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
Cell In [2], line 135
    133 func.py_func(g.dat[1])
    134 print('----------------')
--> 135 func(g.dat[1], True)

File /lan/csv/geomrd3/pankaj/proj/code/pegasus_profile/lib/python/pyenv/lib/python3.10/site-packages/numba/core/dispatcher.py:468, in _DispatcherBase._compile_for_args(self, *args, **kws)
    464         msg = (f"{str(e).rstrip()} \n\nThis error may have been caused "
    465                f"by the following argument(s):\n{args_str}\n")
    466         e.patch_message(msg)
--> 468     error_rewrite(e, 'typing')
    469 except errors.UnsupportedError as e:
    470     # Something unsupported is present in the user code, add help info
    471     error_rewrite(e, 'unsupported_error')

File /lan/csv/geomrd3/pankaj/proj/code/pegasus_profile/lib/python/pyenv/lib/python3.10/site-packages/numba/core/dispatcher.py:409, in _DispatcherBase._compile_for_args.<locals>.error_rewrite(e, issue_type)
    407     raise e
    408 else:
--> 409     raise e.with_traceback(None)

TypingError: Failed in nopython mode pipeline (step: Handle with contexts)
Failed in nopython mode pipeline (step: native lowering)
Failed in nopython mode pipeline (step: Handle with contexts)
Failed in nopython mode pipeline (step: nopython frontend)
Internal error at <numba.core.typeinfer.CallConstraint object at 0x2abc2ec61e70>.
must be builtin_function_or_method
During: resolving callee type: type(ObjModeLiftedWith(<function Data._input__numba_lower__.<locals>.input_lower at 0x2abc2ea9e290>))
During: typing of call at /tmp/ipykernel_67891/2955320401.py (99)

Enable logging at debug level for details.

File "../../../../../tmp/ipykernel_67891/2955320401.py", line 99:
<source missing, REPL/exec in use?>

During: lowering "inp = call $72load_method.1(func=$72load_method.1, args=[], kws=(), vararg=None, varkwarg=None, target=None)" at /tmp/ipykernel_67891/2955320401.py (127)
```

**Comment by agoose77** [comment]
What is `.driver()` in this context?

**Reply by pankajp** [reply]
That was a typo, `driver` was an earlier alias for `input`. I've updated the snippet now, but there's no change in the error.

**Comment by pankajp** [comment]
Changing the implementation to 

```python
    def _input__numba_lower__(context, builder, sig, args):
        print(f'_driver__numba_lower__: sig.args={[type(a) for a in sig.args]}, sig.return_type={type(sig.return_type)}, {len(args)=}')
        ret_type = sig.return_type
        @nb.njit(ret_type(*sig.args))
        def get_data_input(dat):
            with nb.objmode(inp=ret_type):
                inp = dat.input()
            return inp
        def input_lower(dat):
            # Fixme: TypingError: Failed in nopython mode pipeline
            print('get_node() for:', dat.idx)
            return get_data_input(dat)
        return context.compile_internal(builder, input_lower, sig, args)
```

results in another strange numba error:

```
RuntimeError: missing Environment: _ZN08NumbaEnv8__main__4Data21_input__numba_lower__12_3clocals_3e11input_lowerB3v19B42c8tJTC_2fWQI8IW1CiAAYKPM6RBFDjESZRVAJmaQIAEN2ak19RecordViewType_28ak14ArrayView_28ak23RecordArrayType_28_28ak71NumpyArrayType_28array_28int64_2c_201d_2c_20C_29_2c_20_7b_7d_29_2c_20ak71NumpyArrayType_28array_28int64_2c_201d_2c_20C_29_2c_20_7b_7d_29_2c_20ak71NumpyArrayType_28array_28int64_2c_201d_2c_20C_29_2c_20_7b_7d_29_2c_20ak55ListArrayType_28array_28int64_2c_201d_2c_20C_29_2c_20ak262NumpyArrayType_28array_28int64_2c_201d_2c_20C_29_2c_20_7b_7d_29_2c_20_7b_7d_29_29_2c_20_28_27idx_27_2c_20_27input_idx_27_2c_20_27input_port_27_2c_20_27outputs_idx_27_29_2c_20_7b_22__record___22_3a_20_22Data_22_7d_29_2c_20_7b_27Data_27_3a_20_3cclass_20_27__main__136Data_27_3e_2c_20_27graph_27_3a_200_2c_20_28_27__numba_typer___27_2c_20_27Data_27_2c_20_27input_27_2c_20_28_29_29_3a_20_3cfunction_20Data148_input__numba_typer___20at_200x2ba0b8f73490_3e_2c_20_28_27__numba_lower___27_2c_20_27Data_27_2c_20_27input_27_2c_20_28_29_29_3a_20_3cfunction_20Data67_input__numba_lower___20at_200x2ba0b8f73be0_3e_7d_2c_20_28_29_29_29E
```

**Comment by jpivarski** [comment]
Hi!

Unfortunately, I can't tell, by reading the code, whether it's supposed to work or not. Even if it were my own code, I would break it down into smaller pieces and get each piece working before attempting these large code blocks that mix many (probably incompatible) concepts. If one piece doesn't work that seems like it ought to, that would be a bug report.

Awkward Arrays are intended to be used as arguments and return values of functions decorated as `@nb.njit` and its equivalents. These are the high-level, user-facing parts of Numba, as opposed to its mechanism for making extensions. (Awkward is itself a Numba extension.)

Beyond that, if you want to add methods to `ak.Records` nested within the array, you can use the [behavior](https://awkward-array.org/doc/main/reference/ak.behavior.html) argument, but that's an advanced use for making fancy APIs. Before adding functionality as methods, can you compute it as Numba-compiled functions that you just pass the arrays into manually? (Also, it would be best to not put `'Data'` and `'graph'` in the `behavior` dict. That dict is not generic metadata—it's used by Awkward Array for functionality.)

`numba.experimental.structref` defines a record-like structure as a Numba extension. In a sense, `StructRef` is an alternative to Awkward Array, though they differ because Awkward Array stores columnar data and `StructRef` makes conventional "structs." `StructRef` is itself an alternative to `jitclass`, another experimental way to make record-like Numba extensions.

`context.compile_internal` is something you wouldn't ever use unless you're making a Numba extension. It's [not even documented](https://numba.readthedocs.io/en/stable/search.html?q=compile_internal&check_keywords=yes&area=default) in the [Extending Numba documentation](https://numba.readthedocs.io/en/stable/extending/index.html). At this level, you wouldn't be using Awkward Arrays but their lowered equivalents, which aren't documented because they're private implementation details. (We might change them without warning.)

I can't tell what the weakrefs are for, but I doubt they play well with Numba. You probably can't pass them as arguments to a compiled function or as parts of a `StructRef`. Weak references are not listed in [Numba's supported Python features](https://numba.readthedocs.io/en/stable/reference/pysupported.html), and only language features and libraries on that page can be used in a compiled context.

Why `eval('_graph_cache')[idx]`? Why not `_graph_cache[idx]`?

The

```
RuntimeError: missing Environment: _ZN08NumbaEnv8__main__4Data21...
```

looked to me like a mangled C++ name, but I ran it through `c++filt` and it's not a C++ name. It's something that Numba made, and very likely an internal detail.

**Reply by pankajp** [reply]
Hi @jpivarski . Thanks a lot for taking the time to read my post and the elaborate response.
I realize that the code is longer than it should, accumulating cruft over multiple attempts to get it working.

> Beyond that, if you want to add methods to ak.Records nested within the array, you can use the [behavior](https://awkward-array.org/doc/main/reference/ak.behavior.html) argument, but that's an advanced use for making fancy APIs. Before adding functionality as methods, can you compute it as Numba-compiled functions that you just pass the arrays into manually? (Also, it would be best to not put 'Data' and 'graph' in the behavior dict. That dict is not generic metadata—it's used by Awkward Array for functionality.)

I realize `graph` doesn't conform to what awkward array expects and I put it there because I couldn't find any other way to stick in metadata to an awkward array. However, `Data` is exactly intended for the purpose of using a record subclass to add methods to the record (`input()` method here for example). 

> `numba.experimental.structref` defines a record-like structure as a Numba extension. In a sense, StructRef is an alternative to Awkward Array, though they differ because Awkward Array stores columnar data and StructRef makes conventional "structs." StructRef is itself an alternative to jitclass, another experimental way to make record-like Numba extensions.

Yes, an earlier approach used jitclass, but it had more overhead and couldn't cache the compilation. The structref here is intended to store the two awkward arrays.

>`context.compile_internal` is something you wouldn't ever use unless you're making a Numba extension. It's [not even documented](https://numba.readthedocs.io/en/stable/search.html?q=compile_internal&check_keywords=yes&area=default) in the [Extending Numba documentation](https://numba.readthedocs.io/en/stable/extending/index.html). At this level, you wouldn't be using Awkward Arrays but their lowered equivalents, which aren't documented because they're private implementation details. (We might change them without warning.)

Ironically I found `context.compile_internal` in awkward docs ( https://awkward-array.org/doc/main/reference/ak.behavior.html?highlight=compile_internal#overriding-behavior-in-numba ) for the purpose I am using it :)

> I can't tell what the weakrefs are for, but I doubt they play well with Numba. You probably can't pass them as arguments to a compiled function or as parts of a StructRef. Weak references are not listed in [Numba's supported Python features](https://numba.readthedocs.io/en/stable/reference/pysupported.html), and only language features and libraries on that page can be used in a compiled context.

Weakrefs aren't intended to be interfaced with numba or awkward code, but you are right they are extraneous code which isn't necessary here.

`Why eval('_graph_cache')[idx]? Why not _graph_cache[idx]?`

`eval` because otherwise numba wants to pickle and cache the global references as constants and I don't want to pickle and create copies of the full graph and its two awkward arrays.

Thanks again for your inputs, I'll try to simplify the code and make reproduce the problem with less complexity.

**Comment by pankajp** [comment, marked as answer]
I finally got it working.
- The `RuntimeError: missing Environment` got fixed by removing the print() call within the lower() function.
- That resulted in a segfault, which I fixed by adding an access to `.numba_type` so that the numba view is created for the record
  This segfault is what I haven been trying to fix for a long while and which led me to the complex code I shared, changing the graph implementation between a namedtuple, a jitclass, a structmodel and a structref.

```python
    def _input__numba_lower__(context, builder, sig, args):
        print(f'_driver__numba_lower__: sig.args={[type(a) for a in sig.args]}, sig.return_type={type(sig.return_type)}, {len(args)=}')
        ret_type = sig.return_type
        @nb.njit(ret_type(*sig.args))
        def get_data_input(dat):
            with nb.objmode(inp=ret_type):
                inp = dat.input()
                inp.numba_type  # >>> Removing this line causes segfault
            return inp
        def input_lower(dat):
            # Fixme: TypingError: Failed in nopython mode pipeline
            # print('get_node() for:', dat.idx) # >>> having print here causes numba "RuntimeError: missing Environment"
            return get_data_input(dat)
        return context.compile_internal(builder, input_lower, sig, args)
```

**Reply by jpivarski** [reply]
I'm glad this is working!

**Comment by jpivarski** [comment]
I want to keep all of the discussions open. Issues get closed when they're done, but it's valuable to keep discussions around—even if they're resolved—because they're useful to other people with the same questions.
